### DISCUSSION TITLE
Having trouble converting the example app to typescript

### DISCUSSION THREAD

**Discussion by alexkibler** [gladly-team/next-firebase-auth#280] [Q&A]
Hello there!  I'm working on a new app, starting from using the example app to get authentication set up.

I cloned the folder, set up all the firebase stuff, got everything working both locally and in vercel.

Now I'm working on converting everything to use typescript before I start ripping out components and building my own.  However, when I tried converting the `ssr-auth-required.js` component to a `.tsx` file, I got a few errors.  The first were that it didn't like that the `favoriteColor` prop was untyped.  So I created a new interface called Props, typed it, and typed the function component as `FC<Props>` (based on [this stackoverflow response](https://stackoverflow.com/questions/55370851/how-to-fix-binding-element-children-implicitly-has-an-any-type-ts7031).

However, now at the very bottom of the file where we call withAuthUser with the component, it's saying `Argument of type 'FC<Props>' is not assignable to parameter of type 'ComponentType<unknown>'.
  Type 'FunctionComponent<Props>' is not assignable to type 'FunctionComponent<unknown>'.`

Here's the whole file.  The only things I changed from the example are the things I mentioned above, plus I changed `Demo` to `GameTracker`

```
import React, { FC } from 'react'
import {
  useAuthUser,
  withAuthUser,
  withAuthUserTokenSSR,
  AuthAction,
} from 'next-firebase-auth'
import Header from '../components/Header'
import DemoPageLinks from '../components/DemoPageLinks'
import getAbsoluteURL from '../utils/getAbsoluteURL'

const styles = {
  content: {
    padding: 32,
  },
  infoTextContainer: {
    marginBottom: 32,
  },
}

interface Props {
  favoriteColor: string;
}

const GameTracker: FC<Props> = ({ favoriteColor }) => {
  const AuthUser = useAuthUser()
  return (
    <div>
      <Header email={AuthUser.email} signOut={AuthUser.signOut} />
      <div style={styles.content}>
        <div style={styles.infoTextContainer}>
          <h3>Example: SSR + data fetching with ID token</h3>
          <p>
            This page requires authentication. It will do a server-side redirect
            (307) to the login page if the auth cookies are not set.
          </p>
          <p>Your favorite color is: {favoriteColor}</p>
        </div>
        <DemoPageLinks />
      </div>
    </div>
  )
}

export const getServerSideProps = withAuthUserTokenSSR({
  whenUnauthed: AuthAction.REDIRECT_TO_LOGIN,
})(async ({ AuthUser, req }) => {
  // Optionally, get other props.
  // You can return anything you'd normally return from
  // `getServerSideProps`, including redirects.
  // https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering
  const token = await AuthUser.getIdToken()
  const endpoint = getAbsoluteURL('/api/example', req)
  const response = await fetch(endpoint, {
    method: 'GET',
    headers: {
      Authorization: token || 'unauthenticated',
    },
  })
  const data = await response.json()
  if (!response.ok) {
    throw new Error(
      `Data fetching failed with status ${response.status}: ${JSON.stringify(
        data
      )}`
    )
  }
  return {
    props: {
      favoriteColor: data.favoriteColor,
    },
  }
})

export default withAuthUser({
  whenUnauthedAfterInit: AuthAction.REDIRECT_TO_LOGIN,
})(GameTracker)
```

It's probably worth mentioning that I'm an angular developer of 6 years and this is my first attempt at learning to do React development.

The weird thing is, if I _don't_ type the props and just make `GameTracker<FC<any>>`, the error goes away.  But for curiosity's sake, is it possible to fix this with typed props?  Eventually my props will get relatively complicated, so I'd like to type them if I can (but obviously this isn't a blocker for now)

**Comment by gavinharris-dev** [comment]
If you pass in the Type to `withAuthUser` it will then expect a Component with type provided and not Unknown, so you will use:

```
export default withAuthUser<Props>({
  whenUnauthedAfterInit: AuthAction.REDIRECT_TO_LOGIN,
})(GameTracker)
```

Hope that makes sense!
Gav
